[DTensor] Supported `foreach=True` for `clip_grad_norm_`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120919
* __->__ #120910
* #120238

This PR adds support for `clip_grad_norm_(foreach=True)` by implementing `aten._foreach_norm.Scalar` and `aten._foreach_mul_.Tensor`. `foreach=True` is required to get competitive performance with `DTensor`.

`foreach=True` reduces CPU overhead for Llama-7B from 388 ms to 63 ms. Existing flat-parameter FSDP's `clip_grad_norm_` takes 3 ms on CPU 😢 .





cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
[DTensor] Supported `foreach=True` for `clip_grad_norm_`

[ghstack-poisoned]

Commit:
Update on "[DTensor] Supported `foreach=True` for `clip_grad_norm_`"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[DTensor] Supported `foreach=True` for `clip_grad_norm_`"


This PR adds support for `clip_grad_norm_(foreach=True)` by implementing `aten._foreach_norm.Scalar` and `aten._foreach_mul_.Tensor`. `foreach=True` is required to get competitive performance with `DTensor`.







cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm!

Commit:
Update on "[DTensor] Supported `foreach=True` for `clip_grad_norm_`"


This PR adds support for `clip_grad_norm_(foreach=True)` by implementing `aten._foreach_norm.Scalar` and `aten._foreach_mul_.Tensor`. `foreach=True` is required to get competitive performance with `DTensor`.

`foreach=True` reduces CPU overhead for Llama-7B from 388 ms to 63 ms. Existing flat-parameter FSDP's `clip_grad_norm_` takes 3 ms on CPU 😢 .





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

janeyx99 commented on torch/utils/_foreach_utils.py:
```diff
@@ -40,6 +40,4 @@ def _group_tensors_by_device_and_dtype(
-    if device.type not in set(_get_foreach_kernels_supported_devices() + ["cpu"]) or torch.jit.is_scripting():
-        return False
-    return all(t is None or type(t) == torch.Tensor for t in tensors)
+    return device.type in _get_foreach_kernels_supported_devices() + ["cpu"] and not torch.jit.is_scripting()
```
This check should definitely be updated. Could we lift https://github.com/pytorch/pytorch/blob/main/torch/optim/optimizer.py#L47 and incorporate that here? And then DTensor can be dynamically added to the list? Would that work?

Commit:
Update on "[DTensor] Supported `foreach=True` for `clip_grad_norm_`"


This PR adds support for `clip_grad_norm_(foreach=True)` by implementing `aten._foreach_norm.Scalar` and `aten._foreach_mul_.Tensor`. `foreach=True` is required to get competitive performance with `DTensor`.

`foreach=True` reduces CPU overhead for Llama-7B from 388 ms to 63 ms. Existing flat-parameter FSDP's `clip_grad_norm_` takes 3 ms on CPU 😢 .





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

awgu commented on torch/utils/_foreach_utils.py:
```diff
@@ -39,7 +39,9 @@ def _group_tensors_by_device_and_dtype(
 
 
+def _device_has_foreach_support(device: torch.device) -> bool:
+    return device.type in (_get_foreach_kernels_supported_devices() + ["cpu"]) and not torch.jit.is_scripting()
```
I think converting to set takes a linear pass over the list, so we might as well just keep it as a list.

janeyx99 commented on torch/utils/_foreach_utils.py:
```diff
@@ -39,7 +39,9 @@ def _group_tensors_by_device_and_dtype(
-    if device.type not in set(_get_foreach_kernels_supported_devices() + ["cpu"]) or torch.jit.is_scripting():
-        return False
-    return all(t is None or type(t) == torch.Tensor for t in tensors)
+    return _device_has_foreach_support(device) and all(t is None or type(t) == torch.Tensor for t in tensors)
```
Nice, a good followup is to relax the type here to allow for nn.Parameter at the very least. Would be an easy PR 😛

awgu replied:
I think we have to be careful for `nn.Parameter` since we need to know its underlying type (under the `nn.Parameter`). Do you know what is the canonical way to do that?

janeyx99 reviewed (approved):
foreach part lgtm--thanks!

awgu commented:
This test failure exists on [trunk](https://hud.pytorch.org/minihud/pytorch/pytorch/main/2?per_page=50&filter_reruns=true&filter_unstable=true&name_filter=pull%20%2F%20linux-focal-cuda12.1-py3.10-gcc9%20%2F%20test%20(default%2C%205%2C%205%2C%20linux.4xlarge.nvidia.gpu)):
```
python test/test_foreach.py -k test_binary_op_list_error_cases__foreach_add_cuda_bfloat16
```

awgu commented:
@pytorchbot merge -i
